### ISSUE TITLE
SP Flash Tool for Mediatek devices

### ISSUE THREAD

**Issue by drebrez** [postmarketOS/pmbootstrap#920]
The wiko lenny 3 I'm porting right now it's a Mediatek device.
In order to flash the kernel/system to the device it's necessary the proprietary `SP Flash Tool` program.
In the recent versions they added a "console mode" and I'm considering to package it and try to implement the `flasher flash_kernel` and `flasher flash_system` methods.
What do you think about packaging that proprietary software?

**Comment by ollieparanoid** [comment]
Right now we don't execute *any* proprietary binaries on the host while running `pmbootstrap` and I would personally like to keep it that way.

But if there was really no way around it, I think we should at least ask for confirmation before running this proprietary component (related: #756).

I suggest that we write down a **guide in the wiki instead**, and wait until we have a reverse engineered SP Flash Tool available (just like what Heimdall is to Odin). (How is that going @McBitter? Booting a custom kernel + initramfs would be enough actually, we could set up our own flasher from there: https://github.com/postmarketOS/pmbootstrap/issues/456#issuecomment-343686341)

**Comment by drebrez** [comment]
@ollieparanoid What you think about adding the `sp-flash-tool` entry in the `pmb.config.flash_methods` (so that gets prompted during device wizard) and when we run `pmbootstrap flasher *` we show a message with something like this:
```
NOTE: SP Flash Tool is not integrated in pmbootstrap
Please perform the following manual steps:
- open SP Flash Tool
- select scatter file -> aports/device/device-xxxx-yyyy/xxxx-yyyy-scatter.txt
- check boot partition -> $work/chroot_rootfs_xxxx-yyyy/boot/boot.img
- check system partition -> $work/chroot_native/home/pmos/...../system.img
- press Download
- connect the device
```
(side note, in `pmb.config.flash_methods` there is `0xffff` which is also not implemented)

**Comment by ollieparanoid** [comment]
That sounds good! 👍 

You could also note, that convenience symlinks to the files (except for the "scatter" file, is that the partition layout?) can be created with `pmbootstrap export`.

**Comment by drebrez** [comment]
@ollieparanoid yes I was thinking the same about symlinks, maybe even export them automatically . The scatter file defines the partition layout ([example](https://bpaste.net/show/af862b0503a1))

**Comment by ollieparanoid** [comment]
I think it makes sense to store it in the aports folder along with the device package. Because at some point we want to split the `aports` source from the `pmbootstrap` source (#383), and then it's better to be able to update/add such scatter files together with the device packaging instead of the pmbootstrap source.

But it isn't really useful to package it, because we only need it from the Python code in `pmbootstrap` and never on the actual device, right? So maybe just put it in `aports/device/device-xxxx-yyyy/xxxx-yyyy-scatter.txt` and *not* package it. It looks like the file format allows comments, so we could add a comment on top, that links to a wiki page, which explains why this is not inside the package.

**Comment by McBitter** [comment]
For some reason highlighting didn't work. I've pretty much finished with assessing the SP Flashtool. The requirements it places on images themselves are not worth the effort. Based on dump from @drebrez I can say that even flashtool fails to do flashing in some cases.

There might be a work around by skipping some of the setup on the device side. I'm slowly putting together tool for flashing MT6735P based on traffic extracted from SP flash tool and dump from @drebrez. When the POC for MT6735P works, I can try to the same for MT6580M (wiko lenny 3).

**Comment by drebrez** [comment]
@McBitter if you need I can give you some sniffing dump of the flashing operation on wiko lenny 3

**Comment by McBitter** [comment]
Would love to have some USB data in the pcap format. @drebrez

**Comment by ArmsOfSorrow** [comment]
Do you still need help?
I've got a different device, though, an Asus Memo Pad HD7 (ME173X). It has an MT8125 CPU but Android actually identifies it as MT6589, so they're probably very similar.

The SP flash tool version I've used successfully with it is ancient (3.x), though...haven't tried newer ones in fear of bricking it.

**Comment by drebrez** [comment]
@ArmsOfSorrow since SP flash tool is proprietary we don't want to package it inside pmbootstrap, my idea was something like explained in https://github.com/postmarketOS/pmbootstrap/issues/920#issuecomment-345769658, but it's related to the SP Flash tool, instead I was waiting for @McBitter and other people from the #postmarketos-lowlevel channel for a custom solution to flash the mediatek.
@McBitter any good news?
